### ISSUE TITLE
Import Dates 

### ISSUE THREAD

**Issue by beescalante** [cewi/excel#11]
Hi again!

This time I have a problem when trying to import dates from excel.

I have this type of fields in excel with dates '2016-05-12' ready to enter the database. But, the import does not validate the field!

give this error:

> '[errors]' => [
	'birth_date' => [
		'date' => 'The provided value is invalid'
	],
	'entry_date' => [
		'date' => 'The provided value is invalid'
	]
],
'[invalid]' => [
	'birth_date' => (float) 42714,
	'entry_date' => (float) 42623
],

My fields are of type 'Date' for cakephp to validate. What type of field should I use in this case?
Thanks!

**Comment by cewi** [comment]
If you enter dates in Excel, Excel internally stores them as a floats, representing the days since 01.01.1900 (or 01.01.1904 when using excel for mac). PHPExcel seems not to recognize them as  dates and passes floats to the Import-Component. You have two options:

If you store youre dates already in SQL-Date-Format (e.g. '2017-08-15') in your Excel-File, you have to format these fields in Excel as 'text'. The import-component then gets passed them as strings and can validate them as dates.

If you want import all kind of dates from excel, like '15.05.2017' or '08/15/2017', you must format these fields in Excel as dates. Then write a beforeMarshal() function in your Table Object to convert the passed floats to dates. See: https://book.cakephp.org/3.0/en/orm/saving-data.html#modifying-request-data-before-building-entities 
You find al lot of information how to convert excle date folats by serching for 'convert excel date to sql php' 
This is a bit more work but you then are free to use every date format you like.
HTH

**Comment by beescalante** [comment]
Thanks!! It works...
All fields in excel must be text type so that I get the component.

I have another question out about dates and cakephp. Cake does not let me enter older dates of 2012.
Do I have to open that range of dates? 
I hope you can help me!

**Comment by cewi** [comment]
Hi,
glad to hear that it works. The component can import not only strings but also numbers, as integers or floats. For all other types, it may be best to format the Excel cells as text. But as I wrote before, you have the option to convert the imported data as needed in the beforeMarshal() method. dd() the result of the prepareEntityData() method to check how the component imports your data and then adapt the beforeMarshal() method to your needs.

CakePHP can handle dates before 2012! I checked the component too and imported dates like 2001-01-01 or 2006-02-27 without any problems. What are your validation rules?
